### PR TITLE
Fix type comparisons for Nullsafe* functions

### LINKED ISSUE #13600: Bug Report: evalengine null safe comparison and hash functions return wrong results
### Overview of the Issue

Internally in the `evalengine` we use `NullsafeCompare`, `NullsafeHashcode` & `NullsafeHashcode128` to compare values for equality and ordering. These implementations are not currently correct in a number of edge cases. They don't allow for hashing all types like `datetime` and the comparison is wrong for `time` since `time` is not byte ordered. `enum` isn't either (it's ordered by it's underlying enum value, not the string representation). We also don't do the correct casting when different types are compared either.

In other areas in the `evalengine` already have the correct comparison logic that is much more tested. It operates not on `sqltypes.Value` though, but on `eval` types. But for the more complex edge cases, we're probably better off reusing the logic.

### Reproduction Steps

`evalengine.NullsafeCompare(time '101:00:00', time '24:00:00')` returns the wrong result, so any multi-shard query using `time` and ordering would return the wrong results. 

### Binary Version

```sh
All
```

### Operating System and Environment details

```sh
-
```

### Log Fragments

_No response_

### PR DESCRIPTION
**Pull request by dbussink**
This implements additional type support for the Nullsafe* family of  functions. It implements a fast path for common types with equal coercion and then falls back to the generic evalengine logic for all other cases.

## Related Issue(s)

Fixes #13600

## Checklist

-   [x] "Backport to:" labels have been added if this change should be back-ported
-   [x] Tests were added or are not required
-   [x] Did the new or modified tests pass consistently locally and on the CI
-   [x] Documentation was added or is not required

### CODE REVIEW

File: go/mysql/json/weights.go [hunk 1]
```diff
@@ -0,0 +1,153 @@
+package json
```
Comment (systay) [inline go/mysql/json/weights.go:1]: license is missing

File: go/mysql/json/weights_test.go [hunk 2]
```diff
@@ -0,0 +1,28 @@
+package json
```
Comment (systay) [inline go/mysql/json/weights_test.go:1]: license

File: go/vt/vtgate/evalengine/eval.go [hunk 3]
```diff
@@ -226,7 +229,16 @@ func valueToEvalCast(v sqltypes.Value, typ sqltypes.Type) (eval, error) {
 			fval, _ := fastparse.ParseFloat64(v.RawStr())
 			return newEvalFloat(fval), nil
 		default:
-			return nil, vterrors.Errorf(vtrpcpb.Code_INTERNAL, "coercion should not try to coerce this value to a float: %v", v)
+			e, err := valueToEval(v, collations.TypedCollation{
```
Comment (systay) [inline go/vt/vtgate/evalengine/eval.go:232]: we are using this snippet in a few places. worth extracting into a method?

```go
collations.TypedCollation{
  Collation:    collation,
  Coercibility: collations.CoerceCoercible,
  Repertoire:   collations.RepertoireUnicode,
}
```

**Review by systay — APPROVED** [review]
Nice fixes again

File: go/vt/vtgate/evalengine/eval.go (hunk 3 above)
Comment (dbussink) [reply go/vt/vtgate/evalengine/eval.go:232]: @systay that method even exists already! But I'm not using it, fixing it up.

File: go/mysql/json/weights.go [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+	JSON_KEY_TIME        = '\x0A'
+	JSON_KEY_DATETIME    = '\x0B'
+	JSON_KEY_OPAQUE      = '\x0C'
+)
+
+// numericWeightString generates a fixed-width weight string for any JSON
+// number. It requires the `num` representation to be normalized, otherwise
+// the resulting string will not sort.
+func (v *Value) numericWeightString(dst []byte, num string) []byte {
+	const MaxPadLength = 30
```
Comment (frouioui) [inline go/mysql/json/weights.go:47]: nit:
```suggestion
	const maxPadLength = 30
```

File: go/mysql/json/weights.go [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+		// MySQL compat: we follow the same behavior as MySQL does for weight strings in JSON,
+		// where Objects and Arrays are only sorted by their length and not by the values
+		// of their contents.
+		// Note that in MySQL, generating the weight string of a JSON Object or Array will actually
+		// print a warning in the logs! We're not printing anything.
+		dst = append(dst, JSON_KEY_OBJECT)
+		dst = binary.BigEndian.AppendUint32(dst, uint32(v.o.Len()))
+	case TypeArray:
+		dst = append(dst, JSON_KEY_ARRAY)
+		dst = binary.BigEndian.AppendUint32(dst, uint32(len(v.a)))
```
Comment (frouioui) [inline go/mysql/json/weights.go:145]: Why do we always use BigEndian here not also Little Endian?

Comment (dbussink) [reply go/mysql/json/weights.go:145]: It has to be consistent across everything here. And doing it in big endian ensures that it's correctly byte ordered.

**Pull request merged**